active-rfcs/0000-vtu-deprecations.md

```diff
@@ -0,0 +1,188 @@
+# API Deprecation [RFC]
+
+Created: Mar 28, 2020 7:39 PM
+
+- Start Date: 2020.03.28
+- Target Major Version: 1.x
+- Reference Issues:
+- Implementation PR:
+
+# Summary
+
+Deprecate lesser used or bad practice methods and properties from the upcoming VueTestUtils 1.x API, aimed at Vue 3. 
+
+## Motivation
+
+With the current version of Vue Test Utils-beta.x, we notice the API is large and somewhat confusing to most  users, which are not well acquainted with testing Vue apps, leading to bad testing practices. 
+
+This PR aims to trim down the VTU API, replacing listed methods with improved docs, more examples for all the related cases or improved testing methods.
+
+## Detailed design
+
+### Methods
+
+**Wrapper.emittedByOrder -** [Link](https://vue-test-utils.vuejs.org/api/wrapper/#emittedbyorder)
+
+Rarely used, use `emitted` instead.
+
+```js
+expect(wrapper.emitted('change')[0]).toEqual(['param1', 'param2'])
+```
+
+**get -** [Link](https://vue-test-utils.vuejs.org/api/wrapper/#get) 
+
+Recently added, throws error if nothing is matched. Will be merged into `find`.
+
+```js
+expect(wrapper.find('.notExisting')).toThrow()
+
+const element = wrapper.find('.notExisting') // will throw error
+```
+
+**is -** [Link](https://vue-test-utils.vuejs.org/api/wrapper/#is)
+
+Not that much useful. Use `element.tagName` instead. 
+
+```js
+expect(wrapper.element.tagName).toEqual('div')
+```
+
+**isEmpty -** [Link](https://vue-test-utils.vuejs.org/api/wrapper/#isempty)
+
+Use custom matcher like [jest-dom](https://github.com/testing-library/jest-dom#tobeempty) on the element
+
+```js
+expect(wrapper.element).toBeEmpty()
+```
+
+**isVisible** - [Link](https://vue-test-utils.vuejs.org/api/wrapper/#isvisible)
+
+Use custom matcher like [jest-dom](https://github.com/testing-library/jest-dom#tobevisible)
+
+```js
+expect(wrapper.element).toBeVisible()
+```
+
+**isVueInstance -** [Link](https://vue-test-utils.vuejs.org/api/wrapper/#isvueinstance)
+
+Rarely used, no benefits.
+
+**props -** [Link](https://vue-test-utils.vuejs.org/api/wrapper/#props)
+
+Anti-pattern. Test what a prop does, not it's presence or value on the wrapper.
+
+**setData -** [Link](https://vue-test-utils.vuejs.org/api/wrapper/#setdata)
+
+Anti-pattern. Use `data` mounting option to set a specific state.
+
+```js
+const wrapper = mount(Component, { 
+    data () { 
+        return { 
+            field: 'overriden'  
+        }  
+    } 
+})
+``` 
+
+If you really need to update something after it mounts, just use the instance
+
+```js
+wrapper.vm.field = 'updated field'
```

Comment:
I think `setData` makes sense because we can `await setData()` whereas with `wrapper.vm.field = 'updated'` we need to do `await nextTick()` after it

Reply:
What are the use cases for arbitrarily updating a component's data?

Reply:
Use case is usually unit testing when the data is set from something that is mocked so it's impossible to trigger but an integration test is too much for the test in question (could be some displaying or anything)

Also interactions that are too hard/verbose to test with _clicks_ and/or changing _input_'s _values_.

Reply:
Can you give an example of the first case where mocking something makes it impossible to trigger the interaction? I'm not sure I fully understand.

Reply:
A router guard that sets data is one example

Reply:
I still don't understand why you need `setData` for this. Can you provide an example test that shows why `setData` is needed?

Reply:
_I might be playing devil's advocate here, because 99% of times I would suggest against tweaking internal values of a component. But…_

 Imagine a component that does something after clicking a button 100 times. It would make sense to "force" `count` to `99` and then just click once, rather than `trigger('click')` a hundred times.

I guess this is what @posva meant with "interactions that are too hard/verbose to test with clicks". Correct me if I'm wrong 😇 

I start to think that VTU could split their API in two: some basic methods that covered like 85% of common use cases, and then this "advanced" methods (such as `setData`, and even shallow mounting) for edge cases and/or advanced users.

Reply:
a guard can set data inside of it


```js
beforeRouteUpdate(to, from, next) {
  this.things = 'stuff'
  next()
}
```

but we don't necessarily have a way to call that guard

The clicking x times example works as well, others might be related to global events like scrolling, or other things that depend on a device and that are not always easy to mock. When these native calls change the data of the component, you need a way to do it yourself to simulate the native behavior.

Flexibility in VTU is crucial because people write their code in very different ways and we still want them to be able to test it as long as it's not an antipattern, even when they are not using the best practice. This will make the experience with VTU less frustrating because when given a complicated interaction to recreate, it's still possible to write a test for it.

Since we allow setting data directly into a component, `setData` makes sense to make the test writing easier by automatically waiting for one tick.

Reply:
@afontcu for the 100 clicks case....

```
mount(Foo, {
  data() {
    return { count: 99 }
  }
})

wrapper.find('button').click() // 100 times!
```

We will keep support for the data mounting option. This is just referring to `setData` - maybe that wasn't clear. I don't see this as a use case for `setData`

Re @posva (the guard example) that is an interesting example. I still don't really see how `setData` solves this problem - you still have no way to call that hook. Is the assumption in this example that setting `this.things` triggers a watcher to make an API call or something? Can you provide a full example that shows how `setData` helps with router guards?

> Flexibility in VTU is crucial because people write their code in very different ways and we still want them to be able to test it as long as it's not an antipattern, even when they are not using the best practice. 

I agree 100%! VTU should let people test the components they write, regardless of style. That's why I'm interesting in a full example for the router guard you gave above, I don't see how `setData` solves the problem you are describing. 

> global events like scrolling

I think this is outside of the scope of VTU... can jsdom even scroll? I think it's important to remember VTU is not a one stop solution to all your testing needs, in the same way Vue does not solve every front-end problem - it's just one piece of the puzzle. 

> Since we allow setting data directly into a component, setData makes sense to make the test writing easier by automatically waiting for one tick

I'm not sure I understand what you are saying here. Normally you would wait for the next tick with `await nextTick()`?

Reply:
@lmiller1990 About the 100 click cases, you are over simplifying things:
You are assuming the scenario that needs to be tested only needs the 100 value in data but there might be other things that need to happen in a specific order **before** setting the value to 100 to be able to test a scenario. This is where the flexibility of setting the data directly is important. The only point of `setData` is to also return a promise that calls `nextTick` so we can write

```js
await setData({ thing: 'stuff' })
```

instead of 

```js
wrapper.vm.thing = 'stuff'
await nextTick()
```

And this is considering we will still be able to do `wrapper.vm.thing = 'stuff'`

> That's why I'm interesting in a full example for the router

I can't share them, it's something I've come by with many clients. The problem appears whenever the component has a side effect that is outside of it (like changing the url) and that side effect trigger a listener or hook inside the component (like a route guard). That hook changes data **and that data then has an impact on the rendered template**, which is what we assert against. The code inside the guard that matters is setting the data, the rest are side effects like fetching an API. But again, like with the 100 click example, you sometimes need other things to happen before. The problem comes when there is no way to trigger the hook with VTU because it's not a native Vue feature like a watcher.


> I think this is outside of the scope of VTU...

Exactly, and that's when we need changing data, just like the router example. The point is not to be able to recreate the scroll and any other case, because that would be impossible. Instead we go around it because we still need to unit test some of the stuff.

To make my point clear: if we allow directly changing data on the instance (which we should because of the points mentioned above and also it was included in VTU for a reason), we should allow a function like `setData` for it to be consistent with other `set*` functions that return a call to `nextTick`.

I don't understand what in the point above, you are trying to discuss here

Reply:
Thank you for detailed reply @posva . I appreciate your insight.

`setData` (and other methods) should definitely return `nextTick` - that would be very useful.

Regarding what I am trying to discuss: I am trying to discuss the necessity of `setData` (since I did not think it was a useful method). I think some others agreed, that's why we put it in this RFC for a potential deprecation.

Other people seem to think it is an important API. The main case I see is the one you described relating to things jsdom might not handle as well (like scrolling). For this reason and based on this discussion, I think we should *not* deprecate it. I don't see a good way to implement it in Vue 3 yet (I could be missing something), but based on this it seems we need to work something out if this is an important feature to a lot of users.

Reply:
For anyone following this discussion, based on the above comments `setData` will continue to be supported.